### PR TITLE
Use filtered SharedInformerFactories in controller context

### LINKED ISSUE #69: Getting rid of verbose and fragile informer construction
Right now we construct informers through a really horrible and non-type safe SharedInformerFactory, in order to support running in a single namespace (see: https://github.com/jetstack/navigator/blob/master/pkg/controllers/elasticsearch/elasticsearch.go#L311)

We should update this to use a custom generated SharedInformerFactory that supports namespaced resources, and simplify all of this informer handling logic for good.

### PR DESCRIPTION
**Pull request by munnerz**
**What this PR does / why we need it**:

Switches to using the new filtered SharedInformerFactories as part of https://github.com/kubernetes/kubernetes/pull/54660

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

fixes #69 

**Special notes for your reviewer**:

Don't merge until #79 has been merged

**Release note**:

```release-note
NONE
```

/assign @wallrj

### CODE REVIEW

File: pkg/controllers/elasticsearch/elasticsearch.go [hunk 1]
```diff
@@ -311,41 +310,13 @@ func (e *ElasticsearchController) handlePodObject(obj interface{}) {
...
-				metav1.GroupVersionKind{Version: "v1", Kind: "ConfigMap"},
-				coreinformers.NewConfigMapInformer(ctx.Client, ctx.Namespace, time.Second*30, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}),
-			),
+			ctx.SharedInformerFactory.Navigator().V1alpha1().ElasticsearchClusters(),
+			ctx.SharedInformerFactory.Navigator().V1alpha1().Pilots(),
+			ctx.KubeSharedInformerFactory.Apps().V1beta1().StatefulSets(),
+			ctx.KubeSharedInformerFactory.Core().V1().Pods(),
+			ctx.KubeSharedInformerFactory.Core().V1().ServiceAccounts(),
+			ctx.KubeSharedInformerFactory.Core().V1().Services(),
+			ctx.KubeSharedInformerFactory.Core().V1().ConfigMaps(),
```
Comment (wallrj) [inline pkg/controllers/elasticsearch/elasticsearch.go]: 👍 Much neater!

Comment (munnerz) [reply pkg/controllers/elasticsearch/elasticsearch.go]: Our controller constructors are also typed now *at last* :tada: (no more horribly difficult to debug errors!)

File: cmd/controller/app/controller.go [hunk 2]
```diff
@@ -177,3 +182,9 @@ func KubeConfig(apiServerHost string) (*rest.Config, error) {
 
 	return cfg, nil
 }
+
+func NamespaceFilter(namespace string) func(listOptions *metav1.ListOptions) string {
```
Comment (wallrj) [inline cmd/controller/app/controller.go]: ❓ Could you use type `FilterFunc` for the return signature? Is `FilterFunc` public? If not, should it be?

File: pkg/controllers/elasticsearch/elasticsearch.go [hunk 3]
```diff
@@ -69,13 +68,13 @@ type ElasticsearchController struct {
...
-	serviceaccounts cache.SharedIndexInformer,
-	services cache.SharedIndexInformer,
-	configmaps cache.SharedIndexInformer,
+	es navigatorinformers.ElasticsearchClusterInformer,
+	pilots navigatorinformers.PilotInformer,
+	statefulsets appsinformers.StatefulSetInformer,
+	pods coreinformers.PodInformer,
+	serviceaccounts coreinformers.ServiceAccountInformer,
+	services coreinformers.ServiceInformer,
+	configmaps coreinformers.ConfigMapInformer,
```
Comment (wallrj) [inline pkg/controllers/elasticsearch/elasticsearch.go]: ❓ So this is where we get better type checking. Great.

File: third_party/k8s.io/client-go/informers/externalversions/apps/interface.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
+/*
+Copyright 2017 Jetstack Ltd.
```
Comment (wallrj) [inline third_party/k8s.io/client-go/informers/externalversions/apps/interface.go]: ❓ This should be Copyright Kubernetes I guess.

File: cmd/controller/app/controller.go (hunk 2 above)
Comment (munnerz) [reply cmd/controller/app/controller.go]: The issue is that the `FilterFunc` type is generated within each `informer` package, meaning that we'd need to explicitly cast between the two types when we call `NewSharedInformerFactory`.

Ultimately it's a limitation in Go's type system - it doesn't inspect the 'value' side of the type when assessing whether something in fact implements the type, so two types that look identical to you or I, are different according to Go (although *can* be cast, which *does* take account of the 'value' side). There's a good blog post on this somewhere, but can't find it right now. Ping me if you want me to keep digging for it :smile:

File: third_party/k8s.io/client-go/informers/externalversions/apps/interface.go (hunk 4 above)
Comment (munnerz) [reply third_party/k8s.io/client-go/informers/externalversions/apps/interface.go]: I'm not really sure to be honest - it's generated code, so I feel like it's not something that needs to be attributed to the k8s authors. If we were to change *this* to refer to the k8s authors, I feel like we should change *all* code generated by code-generators to be credited to the k8s authors.

**Pull request merged**